libvast/src/logger.cpp

```diff
@@ -15,9 +15,250 @@
+
+    return spdlog::color_mode::never;
+  }();
+
+  auto log_file = caf::get_or(cfg_file, "vast.log-file",
+                              std::string{defaults::logger::log_file});
+
+  auto cmdline_log_file = caf::get_if<std::string>(&cfg_cmd, "vast.log-file");
+  if (cmdline_log_file) {
+    log_file = *cmdline_log_file;
+  }
+
+  if (is_server) {
+    path log_dir = caf::get_or(cfg_file, "vast.db-directory",
+                               defaults::system::db_directory);
+    if (!exists(log_dir)) {
+      if (auto err = mkdir(log_dir)) {
+        // There could be done an on demand spdlogger if wanted,
+        // but this captures the original behaviour (direct write to stderr)
+        std::cerr << "unable to create directory: " << log_dir.str() << ' '
+                  << vast::render(err) << '\n';
+        return false;
+      }
+      log_file = (log_dir / log_file).str();
+    }
+  } else {
+    // Please note, client file does not go to db_directory!
+    // First command line, then config file
+    auto client_log_file
+      = caf::get_if<std::string>(&cfg_cmd, "vast.client-log-file");
+    if (!client_log_file)
+      client_log_file
+        = caf::get_if<std::string>(&cfg_file, "vast.client-log-file");
+    if (client_log_file)
+      log_file = *client_log_file;
+    else // If there is no client log file, turn off file logging
+      vast_file_verbosity = VAST_LOG_LEVEL_QUIET;
+  }
+
+  spdlog::init_thread_pool(8192, 1);
+
+  std::vector<spdlog::sink_ptr> sinks;
+
+  auto stderr_sink
+    = std::make_shared<spdlog::sinks::ansicolor_stderr_sink_mt>(log_color);
+  stderr_sink->set_level(vast_loglevel_to_spd(vast_console_verbosity));
+  auto console_format
+    = caf::get_if<std::string>(&cfg_file, "vast.console-format");
+  if (console_format) {
+    stderr_sink->set_pattern(*console_format);
+  }
+
+  sinks.push_back(stderr_sink);
+
+  if (vast_file_verbosity != VAST_LOG_LEVEL_QUIET) {
+    // When the max file size is reached, close the file, rename it, and create
+    // a new file. Please adopt as needed/
+    // Arguments are,  max file size and the max number of files.
+    auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(
+      log_file, 1024 * 1024 * 10, 3);
+    file_sink->set_level(vast_loglevel_to_spd(vast_file_verbosity));
+    auto file_format = caf::get_if<std::string>(&cfg_file, "vast.file-format");
+    if (file_format) {
+      stderr_sink->set_pattern(*file_format);
+    }
+    sinks.push_back(file_sink);
+  }
+
+  vast_logger = std::make_shared<spdlog::async_logger>(
+    "vast", sinks.begin(), sinks.end(), spdlog::thread_pool(),
+    spdlog::async_overflow_policy::block);
+
+  vast_logger->set_level(vast_loglevel_to_spd(vast_verbosity));
+
+  spdlog::register_logger(vast_logger);
+
+  return true;
+}
+
+void shutdown_spdlog() {
+  VAST_LOG_SPD_DEBUG("shut down logging");
+
+  spdlog::shutdown();
+}
+
+std::shared_ptr<spdlog::logger> logger() {
+  return vast_logger;
+}
```

Comment:
The `shared_ptr`  comes from spdlog, that's it interface

I know that statics in functions are init once only, but they are hard to change once they are initialised.
and since we initialise the logger system in main, before any thread happens (that has actually be the case) that should not matter for the ordering.

But I can look into it if there is potential to improve .

Reply:
I looked at that again. 
There are log messages (trace e.g.) done before the logger gets initialised, and for this there is the dev/null logger.
When the logger setup is called, the instance need to be changed to the real logger.
So this can not be an init once action and then just read the created instance. 
I suggest to discuss the situation next time we meet.

Reply:
Will implement as discussed

Reply:
Implemented as discussed.
Just a technical note
https://en.cppreference.com/w/cpp/language/siof
since the static variable and the function accessing it where in the same translation unit, and no other access to that variable was possible than through that function, SIOF was not an issue in the previous implementation.

The thread safety has not changed, my recommendation is to wait until C++20 , where its possible to pack shared pointer into atomics. But log system should always be fired up before starting threads anyway.